# 1. Eduardo Pérez CHavarría. 
# Proyecto Henry 2. 
# Homicidios

# ETL Homicidios

## 0. Carga

###  Importamos librerias necesarías

Objetivo. 
Identificar para luego homogeneizar:
1. Tipos de datos en columnas.
2. TIpos de datos 
3. Cantidad de nulos


In [1]:
%load_ext autoreload
%autoreload 2
import locale
from datetime import datetime, time
import pandas as pd
import numpy as np
import funciones_propias as fp
fp.limitar


<function funciones_propias.limitar()>

In [2]:
df_homicidios = pd.read_excel("bases/homicidios.xlsx", sheet_name=0)
df_homicidios.head()

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 XY (CABA)         pos x         pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  
3         MOTO-SD    MOTO         SD  
4  MOTO-PASAJEROS    MOTO  PASAJEROS  

[5 rows x 21 columns]

### 0.1. Normalizar los nombres de las columnas

In [3]:
fp.format_nombres(df_homicidios)
df_homicidios.head(1)

Id  N_Victimas      Fecha  Aaaa  Mm  Dd      Hora Hh  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   

                             Lugar_Del_Hecho Tipo_De_Calle  ... Altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...    NaN   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   

                               Dirección_Normalizada Comuna  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   

                               Xy_(Caba)         Pos_X         Pos_Y  \
0  Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   

  Participantes Victima Acusado  
0     MOTO-AUTO    MOTO    AUTO  

[1 rows x 21 columns]

### 0.2. Explorar filas duplicadas

In [4]:
fp.eliminar_duplicadas(df_homicidios)

No hay filas duplicadas.


'nada por eliminar'

### 0.3. Frecuencia de tipos de valor por columna. Incluyendo nulos

In [5]:
fp.tipos_cols(df_homicidios)

Columna                                      Tipo de Valor  \
0                      Id                                      <class 'str'>   
1                      Id                                                NaN   
2              N_Victimas                                      <class 'int'>   
3              N_Victimas                                                NaN   
4                   Fecha  <class 'pandas._libs.tslibs.timestamps.Timesta...   
5                   Fecha                                                NaN   
6                    Aaaa                                      <class 'int'>   
7                    Aaaa                                                NaN   
8                      Mm                                      <class 'int'>   
9                      Mm                                                NaN   
10                     Dd                                      <class 'int'>   
11                     Dd                                                NaN   
12                   Hora                            <class 'datetime.time'>   
13                   Hora                                      <class 'str'>   
14                   Hora                        <class 'datetime.datetime'>   
15                   Hora                                                NaN   
16                     Hh                                      <class 'int'>   
17                     Hh                                      <class 'str'>   
18                     Hh                                                NaN   
19        Lugar_Del_Hecho                                      <class 'str'>   
20        Lugar_Del_Hecho                                                NaN   
21          Tipo_De_Calle                                      <class 'str'>   
22          Tipo_De_Calle                                                NaN   
23                  Calle                                      <class 'str'>   
24                  Calle                                    <class 'float'>   
25                  Calle                                                NaN   
26                 Altura                                    <class 'float'>   
27                 Altura                                                NaN   
28                  Cruce                                      <class 'str'>   
29                  Cruce                                    <class 'float'>   
30                  Cruce                                                NaN   
31  Dirección_Normalizada                                      <class 'str'>   
32  Dirección_Normalizada                                    <class 'float'>   
33  Dirección_Normalizada                                                NaN   
34                 Comuna                                      <class 'int'>   
35                 Comuna                                                NaN   
36              Xy_(Caba)                                      <class 'str'>   
37              Xy_(Caba)                                                NaN   
38                  Pos_X                                      <class 'str'>   
39                  Pos_X                                                NaN   
40                  Pos_Y                                      <class 'str'>   
41                  Pos_Y                                                NaN   
42          Participantes                                      <class 'str'>   
43          Participantes                                                NaN   
44                Victima                                      <class 'str'>   
45                Victima                                                NaN   
46                Acusado                                      <class 'str'>   
47                Acusado                                                NaN   

    Cantidad Más de un tipo de valor (no NaN)  
0        696                               No  
1

Se observa que Hora, Altura, Hh, Calle , Cruce y Dirección_Normalizada tienen varios tipos de datos. Siendo Altura y Cruce los que tienen bastantes datos faltantes. Iremos revisando las columnas una a una. Y al final revisaremos el tema de los datos faltantes juntos. Mientras recuperamos aqui el dato sobre los faltantes:

* Altura: 567 nulos
* Calle: 1 nulo.
* Cruce:  171 nulos
* Dirección Normalizada: 8 nulos


In [6]:
df_filtrado = df_homicidios.loc[df_homicidios["Altura"].isnull() & df_homicidios["Cruce"].isnull()]
df_filtrado

Id  N_Victimas      Fecha  Aaaa  Mm  Dd      Hora  Hh  \
35   2016-0049           1 2016-04-17  2016   4  17  00:00:00   0   
38   2016-0052           1 2016-04-20  2016   4  20  20:00:00  20   
64   2016-0087           1 2016-07-02  2016   7   3  00:10:00   0   
71   2016-0096           1 2016-07-25  2016   7  25  07:00:00   7   
81   2016-0107           1 2016-08-20  2016   8  20  08:22:00   8   
91   2016-0118           1 2016-09-04  2016   9   4  03:30:00   3   
100  2016-0130           1 2016-10-04  2016  10   4  12:30:00  12   
106  2016-0136           1 2016-10-25  2016  10  25  00:00:00   0   
119  2016-0151           1 2016-11-18  2016  11  18  20:35:00  20   
127  2016-0160           1 2016-12-06  2016  12   6  05:30:00   5   
139  2016-0174           1 2016-12-27  2016  12  27  00:00:00   0   
148  2017-0009           1 2017-01-16  2017   1  16  13:56:00  13   
155  2017-0016           1 2017-02-03  2017   2   3  05:12:21   5   
176  2017-0042           1 2017-04-10  2017   4  10  09:00:00   9   
180  2017-0050           2 2017-04-28  2017   4  28  11:08:08  11   
181  2017-0051           1 2017-05-01  2017   5   1  03:47:47   3   
234  2017-0112           2 2017-09-10  2017   9  10  09:32:17   9   
256  2017-0140           1 2017-11-19  2017  11  19  23:22:17  23   
281  2018-0007           1 2018-01-11  2018   1  11  02:30:00   2   
294  2018-0020           1 2018-02-19  2018   2  19  22:00:00  22   
313  2018-0039           1 2018-04-21  2018   4  21  22:15:00  22   
346  2018-0072           1 2018-07-11  2018   7  11  13:15:00  13   
385  2018-0111           1 2018-10-19  2018  10  19  12:13:00  12   
398  2018-0124           1 2018-11-15  2018  11  15  19:40:00  19   
474  2019-0058           1 2019-07-10  2019   7  10  09:20:00   9   
492  2019-0076           1 2019-09-05  2019   9   5  18:00:00  18   
511  2019-0095           1 2019-11-20  2019  11  20  15:30:00  15   
513  2019-0097           1 2019-11-25  2019  11  25  08:50:00   8   
546  2020-0026           1 2020-05-17  2020   5  17  06:40:00   6   
555  2020-0035           1 2020-08-04  2020   8   4  05:30:00   5   
559  2020-0039           1 2020-09-01  2020   9   1  19:17:42  19   
585  2020-0066           1 2020-12-07  2020  12   7  04:30:00   4   
599  2021-0001           1 2021-01-06  2021   1   6  18:00:00  18   
601  2021-0003           1 2021-01-10  2021   1  10  07:40:00   7   
621  2021-0023           1 2021-03-01  2021   3   1  09:20:00   9   
624  2021-0026           1 2021-03-15  2021   3  15  02:35:00   2   
631  2021-0033           1 2021-04-14  2021   4  14  17:50:00  17   
643  2021-0045           1 2021-05-26  2021   5  26  17:24:00  17   
655  2021-0057           1 2021-07-10  2021   7  10  23:30:00  23   
665  2021-0067           1 2021-08-08  2021   8   8  09:29:00   9   
679  2021-0081           1 2021-11-02  2021  11   2  13:35:00  13   
680  2021-0082           1 2021-11-06  2021  11   6  04:39:00   4   

                                       Lugar_Del_Hecho Tipo_De_Calle  ...  \
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....     AUTOPISTA  ...   
38                          AUTOPISTA LUGONES PK 10000     AUTOPISTA  ...   
64    AUTOPISTA 1 SUR PTE ARTURO FRONDIZI Y AV CASEROS     AUTOPISTA  ...   
71        AUTOPISTA DELLEPIANE LUIS TTE. GRAL. KM. 2,3     AUTOPISTA  ...   
81             AUTOPISTA 9 DE JULIO SUR ALT AV MENDOZA     AUTOPISTA  ...   
91          AUTOPISTA 1 SUR PTE ARTURO FRONDIZI KM 2.9     AUTOPISTA  ...   
100               AUTOPISTA 9 DE JULIO SUR Y AV BRASIL     AUTOPISTA  ...   
106                   AU BUENOS AIRES - LA PLATA KM. 4     AUTOPISTA  ...   
119                                                 SD         CALLE  ...   
127                AUTOPISTA PERITO MORENO ALTURA VELZ     AUTOPISTA  ...   
139                               AUTOPISTA 25 DE MAYO     AUTOPISTA  ...   
148                          AU 25 DE MAYO Y AV. BOEDO     AUTOPISTA  ...   
155  AU 25 DE MAYO, KILOMETRO 8.3, SENTIDO 

42 columnas que tienen nulos en conjunto. 
En lo que sigue siempre que haya 696 observaciones la columna ya está sin nulos

## 1. Id

In [7]:
## Esta funcion creada retorna la cantidad de valores unicos por tipo de variable en la columna
fp.unicos_col(df_homicidios["Id"])


Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'str'>     696  [2016-0001, 2016-0002, 2016-0003, 2016-0004, 2...

No se aprecia nada importante que modificar. 

## 2. Numero de victimas

In [8]:
fp.unicos_col(df_homicidios["N_Victimas"])


Cantidad total de observaciones: 696


tipo  conteo    valores
0  <class 'int'>       3  [1, 2, 3]

No parece que haya nada que hacer

## 3. Fecha

In [9]:
fp.unicos_col(df_homicidios["Fecha"])


Cantidad total de observaciones: 696


tipo  conteo  \
0  <class 'pandas._libs.tslibs.timestamps.Timesta...     598   

                                             valores  
0  [2016-01-01 00:00:00, 2016-01-02 00:00:00, 201...

### RECUPERAR PARA EL EDA

In [10]:
fp.frecuencia_unicos(df_homicidios["Fecha"])

Frecuencia de Valores Únicos:
Fecha
2018-12-18    3
2016-11-26    3
2018-12-22    3
2018-02-17    3
2016-02-28    3
             ..
2017-10-30    1
2017-10-27    1
2017-10-26    1
2017-10-17    1
2021-11-18    1
Name: count, Length: 598, dtype: int64


pasamos a date time

In [11]:
df_homicidios['Fecha'] = pd.to_datetime(df_homicidios['Fecha'], format='%Y-%m-%d')


Formato correcto. Nada mas que modificar

## 4. Año

In [12]:
fp.unicos_col(df_homicidios["Aaaa"])


Cantidad total de observaciones: 696


tipo  conteo                               valores
0  <class 'int'>       6  [2016, 2017, 2018, 2019, 2020, 2021]

Todos son int ya. Nada mas que modificar

## 5. Mes

In [13]:
fp.unicos_col(df_homicidios["Mm"])


Cantidad total de observaciones: 696


tipo  conteo                                  valores
0  <class 'int'>      12  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]

Cantidad de meses correctos. Ya estan como int

## 

## 6. Dia

In [14]:
fp.unicos_col(df_homicidios["Dd"])


Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'int'>      31  [1, 2, 3, 10, 21, 24, 29, 8, 14, 15, 17, 28, 4...

Tambien ya como int. Cantidad de dias correcta

## 7. Hora del dia (hh:mm:ss)

In [15]:
fp.unicos_col(df_homicidios["Hora"])


Cantidad total de observaciones: 696


tipo  conteo  \
0      <class 'datetime.time'>     296   
1                <class 'str'>      76   
2  <class 'datetime.datetime'>       3   

                                             valores  
0  [04:00:00, 01:15:00, 07:00:00, 00:00:00, 05:20...  
1  [22:50:00, 16:40:00, 10:12:00, 09:40:00, 22:30...  
2  [1900-01-02 00:00:00, 1900-01-14 07:12:00, 190...

vamos a ver los valores str

In [16]:
valores_str = df_homicidios["Hora"].loc[df_homicidios["Hora"].apply(lambda x: isinstance(x, str))]

print(valores_str.unique())

['22:50:00' '16:40:00' '10:12:00' '09:40:00' '22:30:00' '21:15:00'
 '02:30:00' '17:50:00' '10:30:00' '13:30:00' '08:15:00' '20:00:00'
 '01:00:00' '10:15:00' '04:15:00' '07:15:00' '12:30:00' '22:00:00'
 '10:45:00' '00:00:00' '06:00:00' '23:00:00' '04:45:00' '15:45:00'
 '18:00:00' '01:30:00' '13:00:00' '18:15:00' '16:45:00' '23:15:00'
 '06:30:00' '09:00:00' '07:00:00' '03:15:00' '18:45:00' '20:30:00'
 '11:15:00' '19:20:00' '17:56:00' '10:34:00' '08:20:00' '08:41:00'
 '05:35:00' '00:59:00' '22:20:00' '05:00:00' '14:25:00' '00:15:00'
 '00:30:00' '17:51:00' '18:12:00' '22:02:00' '14:20:00' '22:12:00'
 '03:05:00' '12:13:00' '09:18:00' '11:00:00' '02:15:00' '14:49:00'
 '13:34:00' '16:32:00' '03:32:00' '19:40:00' '21:12:00' '09:14:00'
 '04:30:00' '04:01:00' '03:00:00' '08:26:00' '14:07:00' '21:00:00'
 '15:42:00' '09:30:00' '05:25:00' 'SD']


se aprecia que hay un sd y los demas son fechas pero en formato cadena. Vamos a convertirlos en datetime.time

In [17]:
df_homicidios["Hora"] = df_homicidios["Hora"].apply(lambda x: np.nan if pd.isnull(x) else x.time() if isinstance(x, datetime) else (x if isinstance(x, time) else (datetime.strptime(x, '%H:%M:%S').time() if isinstance(x, str) and ':' in x else np.nan)))
#Veamos de nuevo frecuencia de valores unicos
fp.unicos_col(df_homicidios["Hora"])


Cantidad total de observaciones: 696


tipo  conteo  \
0  <class 'datetime.time'>     322   
1          <class 'float'>       1   

                                             valores  
0  [04:00:00, 01:15:00, 07:00:00, 00:00:00, 05:20...  
1                                              [nan]

Vamos a imputar la media al valor faltante

In [18]:
# Valor específico para llenar los faltantes
valor_especifico = datetime.strptime('1:11:11', '%H:%M:%S').time()

# Rellenar los valores faltantes con el valor específico
df_homicidios["Hora"] = df_homicidios["Hora"].fillna(valor_especifico)

In [19]:
from datetime import datetime, timedelta

# Pasar "Hora" a minutos desde la medianoche
df_homicidios["Hora_minutos"] = df_homicidios["Hora"].apply(lambda x: x.hour * 60 + x.minute)

# Calcular la media de los valores existentes en la columna "Hora" en minutos,  
media_minutos = df_homicidios.loc[df_homicidios["Hora"] != valor_especifico, "Hora_minutos"].mean()

# Convertir la media de minutos a formato de tiempo
media_hora = datetime.min + timedelta(minutes=media_minutos)

# Asegurarse de que la media_hora tenga la precisión adecuada (fracciones de segundo)
media_hora = media_hora.replace(microsecond=0)

# Rellenar los valores faltantes con la media recien calculada
df_homicidios["Hora"] = df_homicidios["Hora"].apply(lambda x: media_hora.time() if x == valor_especifico else x)

# Eliminar la columna temporal "Hora_minutos"
df_homicidios = df_homicidios.drop(columns=["Hora_minutos"])


In [20]:
fp.frecuencia_unicos(df_homicidios["Hora"])

Frecuencia de Valores Únicos:
Hora
16:00:00    10
21:00:00    10
09:00:00    10
12:30:00    10
14:00:00    10
            ..
10:12:00     1
00:10:00     1
05:10:00     1
16:10:00     1
00:43:00     1
Name: count, Length: 323, dtype: int64


convertimos a formato time. 

In [21]:
df_homicidios['Hora'] = df_homicidios['Hora'].apply(lambda x: datetime.strptime(str(x), '%H:%M:%S').time() if pd.notnull(x) else None)



comprobamos el tipo de objeto 

In [22]:
tipo_de_objeto = type(df_homicidios['Hora'][0])
print("Tipo de objeto de la columna 'Hora':", tipo_de_objeto)


Tipo de objeto de la columna 'Hora': <class 'datetime.time'>



Formato ya es correcto. Se cambiaron las fechas que incluian registro de año. Y Las cadenas fueron reemplazadas. 

## 8 Hora (unicamente)

In [23]:
fp.unicos_col(df_homicidios["Hh"])


Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'int'>      24  [4, 1, 7, 0, 5, 18, 19, 15, 11, 22, 16, 9, 23,...
1  <class 'str'>       1                                               [SD]

imputamos el valor a sd


In [24]:
media_minutos/60

12.022541966426859

In [25]:
df_homicidios["Hh"] = df_homicidios["Hh"].replace("SD", "12")


coercionamos a int de nuevo para que la cadena se convierta a int

In [26]:
df_homicidios["Hh"] = pd.to_numeric(df_homicidios["Hh"], errors='coerce').astype('Int64')


In [27]:
fp.unicos_col(df_homicidios["Hh"])


Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'int'>      24  [4, 1, 7, 0, 5, 18, 19, 15, 11, 22, 16, 9, 23,...

Ya todos son INT sin valores perdidos

## 9. Lugar del hecho (direccion)

In [28]:
fp.unicos_col(df_homicidios["Lugar_Del_Hecho"])


Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'str'>     683  [AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ, AV...

Exploramos los valores unicos que hay

In [29]:
# Función propia creada, quita el limite a la respuesta de la celda
fp.sin_limites()

# Llamar a la función
output_df = fp.frecuencia_unicos(df_homicidios["Lugar_Del_Hecho"])
output_df
# Funciòn creada, restaura la configuración de visualización predeterminada
fp.limitar()

Frecuencia de Valores Únicos:
Lugar_Del_Hecho
AV 27 DE FEBRERO Y AV ESCALADA                                                           4
CASTILLO, RAMON S., PRES. AV. Y CALLE 12                                                 2
PAZ, GRAL. AV. Y DEL LIBERTADOR AV.                                                      2
AV LA PLATA Y CNEL GREGORIO POMAR                                                        2
CANTILO, INT. Y UDAONDO, GUILLERMO AV.                                                   2
Rivadavia Av. y Pedernera                                                                2
AV AMANCIO ALCORTA Y BONAVENA                                                            2
AV. INDEPENDENCIA Y VIRREY CEVALLOS                                                      2
Nazca Av. y Rivadavia Av.                                                                2
SAN PEDRITO AV. Y DIRECTORIO AV.                                                         2
CORDOBA AV. Y MADERO, EDUARDO AV.           

mas adelante existe una direccion normalizada. Por lo pronto aqui no hay mas que hacer, salvo notar que GRAL PAZ es avenida

## 

## 10. Tipo de calle

In [30]:
fp.unicos_col(df_homicidios["Tipo_De_Calle"])



Cantidad total de observaciones: 696


tipo  conteo                                valores
0  <class 'str'>       4  [AVENIDA, GRAL PAZ, CALLE, AUTOPISTA]

In [31]:
fp.frecuencia_unicos(df_homicidios["Tipo_De_Calle"])


Frecuencia de Valores Únicos:
Tipo_De_Calle
AVENIDA      429
CALLE        136
AUTOPISTA     66
GRAL PAZ      65
Name: count, dtype: int64


Por alguna razon tipo de calle tiene a GRAL, PAZ, como un tipo, cuando es una avenida. Es necesario convertir ese valor a AVENIDA

In [32]:
df_homicidios["Tipo_De_Calle"] = df_homicidios["Tipo_De_Calle"].replace("GRAL PAZ", "AVENIDA")
fp.frecuencia_unicos(df_homicidios["Tipo_De_Calle"])

Frecuencia de Valores Únicos:
Tipo_De_Calle
AVENIDA      494
CALLE        136
AUTOPISTA     66
Name: count, dtype: int64


## 

## 11. Calle 

In [33]:
fp.unicos_col(df_homicidios["Calle"])



Cantidad total de observaciones: 696


tipo  conteo                                            valores
0    <class 'str'>     279  [PIEDRA BUENA AV., PAZ, GRAL. AV., ENTRE RIOS ...
1  <class 'float'>       1                                              [nan]

In [34]:

fp.frecuencia_unicos(df_homicidios["Calle"])

Frecuencia de Valores Únicos:
Calle
PAZ, GRAL. AV.                                57
RIVADAVIA AV.                                 19
DEL LIBERTADOR AV.                            18
AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI    14
ALBERDI, JUAN BAUTISTA AV.                    13
                                              ..
PILCOMAYO                                      1
CERVANTES                                      1
CONGRESO AV.                                   1
AUSTRALIA AV.                                  1
PADRE CARLOS MUJICA                            1
Name: count, Length: 279, dtype: int64


intentamos normalizarla

In [35]:
df_homicidios['Calle_trans']=fp.transformar_calle(df_homicidios["Calle"])
# Imprimir las columnas 'Cruce' y 'Cruce_transformado'
df_homicidios[['Calle', 'Calle_trans']].head(5)

Calle       Calle_trans
0  PIEDRA BUENA AV.  Av. Buena Piedra
1    PAZ, GRAL. AV.    Av. Gral. Paz,
2    ENTRE RIOS AV.    Av. Rios Entre
3    LARRAZABAL AV.    Av. Larrazabal
4      SAN JUAN AV.      Av. Juan San

In [36]:
df_homicidios = df_homicidios.drop('Calle', axis=1)

## 12. Altura

In [37]:
fp.unicos_col(df_homicidios["Altura"])



Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'float'>     127  [nan, 2034.0, 1366.0, 156.0, 1483.0, 30.0, 147...

In [38]:
fp.frecuencia_unicos(df_homicidios["Altura"])


Frecuencia de Valores Únicos:
Altura
901.0     2
4300.0    2
365.0     2
1893.0    1
466.0     1
         ..
5750.0    1
2135.0    1
5400.0    1
2114.0    1
709.0     1
Name: count, Length: 126, dtype: int64


Se puede observar en la longitud que unicamente hay 126 valores con informacion. En otra columna están las coordenadas, las utilizaremos para reemplazar después esta colummna Por lo pronto la eliminaremos

In [39]:
df_homicidios = df_homicidios.drop('Altura', axis=1)


## 13 Cruce

In [40]:
fp.unicos_col(df_homicidios["Cruce"])



Cantidad total de observaciones: 696


tipo  conteo                                            valores
0    <class 'str'>     317  [FERNANDEZ DE LA CRUZ, F., GRAL. AV., DE LOS C...
1  <class 'float'>       1                                              [nan]

In [41]:
fp.sin_limites()
fp.frecuencia_unicos(df_homicidios["Cruce"])
fp.limitar()


Frecuencia de Valores Únicos:
Cruce
PAZ, GRAL. AV.                                14
ESCALADA AV.                                   9
RIVADAVIA AV.                                  8
INDEPENDENCIA AV.                              7
BALBIN, RICARDO, DR. AV.                       6
DEL LIBERTADOR AV.                             6
PUEYRREDON AV.                                 6
FERNANDEZ DE LA CRUZ, F., GRAL. AV.            5
9 DE JULIO AV.                                 5
PERON, EVA AV.                                 5
RIVERA INDARTE AV.                             5
ROCA, CNEL. AV.                                5
ALBERDI, JUAN BAUTISTA AV.                     5
VARELA AV.                                     4
ALEM, LEANDRO N. AV.                           4
CALLAO AV.                                     4
BRASIL                                         4
LA PAMPA                                       4
RABANAL, FRANCISCO, INTENDENTE AV.             4
PARAGUAY                         

intentamos normalizarla

In [42]:
df_homicidios['Cruce_trans']=fp.transformar_calle(df_homicidios["Cruce"])
# Imprimir las columnas 'Cruce' y 'Cruce_transformado'
df_homicidios[['Cruce', 'Cruce_trans']].head(10)


Cruce                          Cruce_trans
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.  Av. Gral. F., Cruz, La De Fernandez
1                  DE LOS CORRALES AV.                  Av. Corrales Los De
2                                  NaN                                     
3             VILLEGAS, CONRADO, GRAL.             Gral. Conrado, Villegas,
4              SAENZ PE?A, LUIS, PRES.              Pres. Luis, Peña, Saenz
5                         ESCALADA AV.                         Av. Escalada
6                 GONZALEZ, JOAQUIN V.                 V. Joaquin Gonzalez,
7                  DE LOS CORRALES AV.                  Av. Corrales Los De
8                IRIGOYEN, BERNARDO DE                De Bernardo Irigoyen,
9                                  NaN

In [43]:
df_homicidios = df_homicidios.drop('Cruce', axis=1)

Como notamos, solo hay valores aquí cuando no había valores en altura, es decir cuando los crimenes ocurren entre calles. Por ello cambiaremos los nulos a Nombre de la calle"+"sin cruce"

In [44]:
# Reemplazar nulos en "Cruce_trans" con el valor de "Calle_trans" + " sin cruce"
df_homicidios['Cruce_trans'] = df_homicidios.apply(
    lambda row: row['Calle_trans'] + ' Sin cruce' if pd.isna(row['Cruce_trans']) else row['Cruce_trans'],
    axis=1
)



## 14 Dirección normalizada

In [45]:
fp.unicos_col(df_homicidios["Dirección_Normalizada"])



Cantidad total de observaciones: 696


tipo  conteo                                            valores
0    <class 'str'>     635  [PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., ...
1  <class 'float'>       1                                              [nan]

Cambiar Na a "SIn dato"

In [46]:
df_homicidios["Dirección_Normalizada"] = df_homicidios["Dirección_Normalizada"].fillna("Sin dato")


Todo correcto hasta el momento

## 15 COmuna

In [47]:
fp.unicos_col(df_homicidios["Comuna"])



Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'int'>      16  [8, 9, 1, 11, 15, 4, 7, 12, 3, 13, 14, 10, 6, ...

In [48]:
# Diccionario con el mapeo de comuna a barrios
comunas_barrios = {
    1: ['Retiro', 'San Nicolás', 'Puerto Madero', 'San Telmo', 'Montserrat', 'Constitución'],
    2: ['Recoleta'],
    3: ['Balvanera', 'San Cristóbal'],
    4: ['La Boca', 'Barracas', 'Parque Patricios', 'Nueva Pompeya'],
    5: ['Almagro', 'Boedo'],
    6: ['Caballito'],
    7: ['Flores', 'Parque Chacabuco'],
    8: ['Villa Soldati', 'Villa Riachuelo', 'Villa Lugano'],
    9: ['Liniers', 'Mataderos', 'Parque Avellaneda'],
    10: ['Villa Real', 'Monte Castro', 'Versalles', 'Floresta', 'Vélez Sarsfield', 'Villa Luro'],
    11: ['Villa General Mitre', 'Villa Devoto', 'Villa del Parque', 'Villa Santa Rita'],
    12: ['Coghlan', 'Saavedra', 'Villa Urquiza', 'Villa Pueyrredón'],
    13: ['Núñez', 'Belgrano', 'Colegiales'],
    14: ['Palermo'],
    15: ['Chacarita', 'Villa Crespo', 'La Paternal', 'Villa Ortúzar', 'Agronomía', 'Parque Chas']
}

# Añadir la nueva columna "Barrios" al DataFrame df_siniestros
df_homicidios['Barrios'] = df_homicidios['Comuna'].map(comunas_barrios)



Todo bien, después obtendremos los datos más especificos sobre la comuna

## 16 X y Y caba

In [49]:
fp.unicos_col(df_homicidios["Xy_(Caba)"])



Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'str'>     606  [Point (98896.78238426 93532.43437792), Point ...

In [50]:
df_homicidios["Xy_(Caba)"]

0        Point (98896.78238426 93532.43437792)
1        Point (95832.05571093 95505.41641999)
2       Point (106684.29090040 99706.57687843)
3        Point (99840.65224780 94269.16534422)
4      Point (106980.32827929 100752.16915795)
                        ...                   
691     Point (102728.60090138 98186.24929177)
692      Point (99624.29795829 97569.69801131)
693     Point (99116.45492358 101045.23284826)
694      Point (99324.54463985 97676.26932409)
695    Point (107664.16647795 104708.63962087)
Name: Xy_(Caba), Length: 696, dtype: object

vamos a extraerlos por si hubiera que trabajar con ellos

In [51]:
import re

# Aplicar la función lambda a la columna 'Xy_(Caba)'
df_homicidios[['X_Latitud', 'Y_Longitud']] = df_homicidios['Xy_(Caba)'].apply(
    lambda coordenada: pd.Series(
        [float(coord) if coord != '.' else None for coord in re.sub(r'Point|\(|\)', '', coordenada).split()]
    )
)

# Muestra el DataFrame con las nuevas columnas de Latitud y Longitud
df_homicidios[['Xy_(Caba)', 'X_Latitud', 'Y_Longitud']]



Xy_(Caba)      X_Latitud     Y_Longitud
0      Point (98896.78238426 93532.43437792)   98896.782384   93532.434378
1      Point (95832.05571093 95505.41641999)   95832.055711   95505.416420
2     Point (106684.29090040 99706.57687843)  106684.290900   99706.576878
3      Point (99840.65224780 94269.16534422)   99840.652248   94269.165344
4    Point (106980.32827929 100752.16915795)  106980.328279  100752.169158
..                                       ...            ...            ...
691   Point (102728.60090138 98186.24929177)  102728.600901   98186.249292
692    Point (99624.29795829 97569.69801131)   99624.297958   97569.698011
693   Point (99116.45492358 101045.23284826)   99116.454924  101045.232848
694    Point (99324.54463985 97676.26932409)   99324.544640   97676.269324
695  Point (107664.16647795 104708.63962087)  107664.166478  104708.639621

[696 rows x 3 columns]

In [52]:
df_homicidios = df_homicidios.drop('Xy_(Caba)', axis=1)


In [53]:
# Mostrar información sobre tipos de datos y conteo de valores
info_df = df_homicidios[['X_Latitud', 'Y_Longitud']].info()

# Conteo de valores únicos en cada columna
unique_counts = df_homicidios[['X_Latitud', 'Y_Longitud']].nunique()

print(info_df)
print(unique_counts)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   X_Latitud   682 non-null    float64
 1   Y_Longitud  682 non-null    float64
dtypes: float64(2)
memory usage: 11.0 KB
None
X_Latitud     605
Y_Longitud    605
dtype: int64


En realidad más adelante tenemos latitud y longitud, por ende eliminamos aquí las columnas de coordenadas caba. Se deja la limpieza solo para ver cómo se trabajó


In [54]:
df_homicidios = df_homicidios.drop(['X_Latitud', 'Y_Longitud'], axis=1)


## 17 Latitud y longitud (Pos_x y POX_Y)

In [55]:
fp.unicos_col(df_homicidios["Pos_X"])



Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'str'>     605  [-58.47533969, -58.50877521, -58.39040293, -58...

In [56]:
fp.unicos_col(df_homicidios["Pos_Y"])


Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'str'>     605  [-34.68757022, -34.66977709, -34.63189362, -34...

In [57]:
fp.unicos_col(df_homicidios["Pos_X"])


Cantidad total de observaciones: 696


tipo  conteo                                            valores
0  <class 'str'>     605  [-58.47533969, -58.50877521, -58.39040293, -58...

In [58]:
fp.frecuencia_unicos(df_homicidios['Pos_Y'])


Frecuencia de Valores Únicos:
Pos_Y
.               12
-34.68475866     5
-34.54979510     4
-34.53476874     4
-34.66977709     4
                ..
-34.58498660     1
-34.59573848     1
-34.63997430     1
-34.64234110     1
-34.58679619     1
Name: count, Length: 605, dtype: int64


In [59]:
df_homicidios[['Pos_X', 'Pos_Y']] = df_homicidios[['Pos_X', 'Pos_Y']].replace('[^0-9.,-]', np.nan, regex=True)
df_homicidios[['Pos_X', 'Pos_Y']] = df_homicidios[['Pos_X', 'Pos_Y']].replace('.', np.nan)
# Convertir las cadenas a tipo numérico (float)
df_homicidios[['Pos_X', 'Pos_Y']] = df_homicidios[['Pos_X', 'Pos_Y']].apply(lambda x: pd.to_numeric(x.str.replace(',', '.'), errors='coerce'))



In [60]:
fp.frecuencia_unicos(df_homicidios['Pos_X'])


Frecuencia de Valores Únicos:
Pos_X
-58.444513    5
-58.500738    4
-58.467435    4
-58.508775    4
-58.469640    3
             ..
-58.391006    1
-58.376977    1
-58.458074    1
-58.420239    1
-58.379762    1
Name: count, Length: 604, dtype: int64


In [61]:
fp.frecuencia_unicos(df_homicidios['Pos_Y'])

Frecuencia de Valores Únicos:
Pos_Y
-34.684759    5
-34.549795    4
-34.534769    4
-34.669777    4
-34.630706    3
             ..
-34.584987    1
-34.595738    1
-34.639974    1
-34.642341    1
-34.586796    1
Name: count, Length: 604, dtype: int64


los valores faltantes los vamos a dropear ya que solo son 2

In [62]:
df_homicidios = df_homicidios.dropna(subset=['Pos_X', 'Pos_Y'])


In [63]:
# Aplicar la función al dataframe df_homicidios
#df_homicidios = fp.convertir_coordenadas(df_homicidios[['Pos_X', 'Pos_Y']])


## 18 Participantes

In [64]:
fp.unicos_col(df_homicidios["Participantes"])



Cantidad total de observaciones: 684


tipo  conteo                                            valores
0  <class 'str'>      41  [MOTO-AUTO, AUTO-PASAJEROS, MOTO-SD, MOTO-PASA...

In [65]:
fp.frecuencia_unicos(df_homicidios["Participantes"])

Frecuencia de Valores Únicos:
Participantes
PEATON-PASAJEROS       105
MOTO-AUTO               83
PEATON-AUTO             76
MOTO-CARGAS             74
MOTO-PASAJEROS          45
MOTO-OBJETO FIJO        39
PEATON-CARGAS           37
PEATON-MOTO             30
AUTO-AUTO               30
MOTO-MOTO               25
AUTO-OBJETO FIJO        22
MULTIPLE                17
AUTO-CARGAS             14
BICICLETA-CARGAS        10
AUTO-PASAJEROS           9
BICICLETA-PASAJEROS      8
BICICLETA-AUTO           8
MOTO-SD                  8
PEATON-BICICLETA         5
SD-SD                    4
AUTO-SD                  4
CARGAS-CARGAS            4
PEATON-SD                3
PASAJEROS-PASAJEROS      3
MOTO-OTRO                2
MOTO-BICICLETA           2
SD-AUTO                  2
MOTO-MOVIL               2
CARGAS-OBJETO FIJO       1
CARGAS-AUTO              1
MOVIL-PASAJEROS          1
SD-CARGAS                1
SD-MOTO                  1
MOVIL-CARGAS             1
PASAJEROS-AUTO           1
AUTO-MOVIL 

Normalizamos nombres

In [66]:
df_homicidios["Participantes"] = df_homicidios["Participantes"].apply(lambda x: '-'.join(word.title() for word in x.split('-')))

# Mostrar el DataFrame actualizado
print(df_homicidios["Participantes"])


0           Moto-Auto
1      Auto-Pasajeros
2           Moto-Auto
3             Moto-Sd
4      Moto-Pasajeros
            ...      
691         Moto-Auto
692         Moto-Auto
693       Moto-Cargas
694       Auto-Cargas
695    Bicicleta-Auto
Name: Participantes, Length: 684, dtype: object


No parece haber problemas con esta columna

## 19 Victima

In [67]:
fp.unicos_col(df_homicidios["Victima"])



Cantidad total de observaciones: 684


tipo  conteo                                            valores
0  <class 'str'>      10  [MOTO, AUTO, PEATON, SD, CARGAS, BICICLETA, PA...

In [68]:
fp.frecuencia_unicos(df_homicidios["Victima"])

Frecuencia de Valores Únicos:
Victima
MOTO           288
PEATON         261
AUTO            82
BICICLETA       29
SD               8
CARGAS           7
PASAJEROS        5
MOVIL            2
OBJETO FIJO      1
PEATON_MOTO      1
Name: count, dtype: int64


Normalizamos

In [69]:
df_homicidios["Victima"] = df_homicidios["Victima"].apply(lambda x: '-'.join(word.title() for word in x.split('-')))

# Mostrar el DataFrame actualizado
print(df_homicidios["Victima"])


0           Moto
1           Auto
2           Moto
3           Moto
4           Moto
         ...    
691         Moto
692         Moto
693         Moto
694         Auto
695    Bicicleta
Name: Victima, Length: 684, dtype: object


## 20 Acusado

In [70]:
fp.unicos_col(df_homicidios["Acusado"])


Cantidad total de observaciones: 684


tipo  conteo                                            valores
0  <class 'str'>      10  [AUTO, PASAJEROS, SD, OBJETO FIJO, CARGAS, MOT...

In [71]:
fp.frecuencia_unicos(df_homicidios["Acusado"])

Frecuencia de Valores Únicos:
Acusado
AUTO           202
PASAJEROS      172
CARGAS         141
OBJETO FIJO     61
MOTO            57
SD              20
MULTIPLE        17
BICICLETA        7
OTRO             6
TREN             1
Name: count, dtype: int64


normalizamos

In [72]:
df_homicidios["Acusado"] = df_homicidios["Acusado"].apply(lambda x: '-'.join(word.title() for word in x.split('-')))

# Mostrar el DataFrame actualizado
print(df_homicidios["Acusado"])


0           Auto
1      Pasajeros
2           Auto
3             Sd
4      Pasajeros
         ...    
691         Auto
692         Auto
693       Cargas
694       Cargas
695         Auto
Name: Acusado, Length: 684, dtype: object


## 22. Columnas finales

Revisamos el resultado de las columnas

In [73]:
fp.tipos_cols(df_homicidios)

Columna                                      Tipo de Valor  \
0                      Id                                      <class 'str'>   
1                      Id                                                NaN   
2              N_Victimas                                      <class 'int'>   
3              N_Victimas                                                NaN   
4                   Fecha  <class 'pandas._libs.tslibs.timestamps.Timesta...   
5                   Fecha                                                NaN   
6                    Aaaa                                      <class 'int'>   
7                    Aaaa                                                NaN   
8                      Mm                                      <class 'int'>   
9                      Mm                                                NaN   
10                     Dd                                      <class 'int'>   
11                     Dd                                                NaN   
12                   Hora                            <class 'datetime.time'>   
13                   Hora                                                NaN   
14                     Hh                                      <class 'int'>   
15                     Hh                                                NaN   
16        Lugar_Del_Hecho                                      <class 'str'>   
17        Lugar_Del_Hecho                                                NaN   
18          Tipo_De_Calle                                      <class 'str'>   
19          Tipo_De_Calle                                                NaN   
20  Dirección_Normalizada                                      <class 'str'>   
21  Dirección_Normalizada                                                NaN   
22                 Comuna                                      <class 'int'>   
23                 Comuna                                                NaN   
24                  Pos_X                                    <class 'float'>   
25                  Pos_X                                                NaN   
26                  Pos_Y                                    <class 'float'>   
27                  Pos_Y                                                NaN   
28          Participantes                                      <class 'str'>   
29          Participantes                                                NaN   
30                Victima                                      <class 'str'>   
31                Victima                                                NaN   
32                Acusado                                      <class 'str'>   
33                Acusado                                                NaN   
34            Calle_trans                                      <class 'str'>   
35            Calle_trans                                                NaN   
36            Cruce_trans                                      <class 'str'>   
37            Cruce_trans                                                NaN   
38                Barrios                                     <class 'list'>   
39                Barrios                                                NaN   

    Cantidad Más de un tipo de valor (no NaN)  
0        684                               No  
1          0                               No  
2        684                               No  
3          0                               No  
4        684                               No  
5          0                               No  
6        684                               No  
7          0                               No  
8        684                               No  
9          0                               No  
10       684                               No  
11         0                               No  
12       684                               No  
13         0                               No  
14       684     

In [74]:
df_homicidios.info()

<class 'pandas.core.frame.DataFrame'>
Index: 684 entries, 0 to 695
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Id                     684 non-null    object        
 1   N_Victimas             684 non-null    int64         
 2   Fecha                  684 non-null    datetime64[ns]
 3   Aaaa                   684 non-null    int64         
 4   Mm                     684 non-null    int64         
 5   Dd                     684 non-null    int64         
 6   Hora                   684 non-null    object        
 7   Hh                     684 non-null    Int64         
 8   Lugar_Del_Hecho        684 non-null    object        
 9   Tipo_De_Calle          684 non-null    object        
 10  Dirección_Normalizada  684 non-null    object        
 11  Comuna                 684 non-null    int64         
 12  Pos_X                  684 non-null    float64       
 13  Pos_Y     

Eliminamos columnas que tienen información redundante

In [75]:
eliminar = ['Aaaa', 'Mm', 'Dd', 'Hh', 'Dirección_Normalizada', 'Participantes', 'Calle_trans', "Cruce_trans"]

df_homicidios.drop(columns=eliminar, inplace=True)


In [76]:
df_homicidios.info()

<class 'pandas.core.frame.DataFrame'>
Index: 684 entries, 0 to 695
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Id               684 non-null    object        
 1   N_Victimas       684 non-null    int64         
 2   Fecha            684 non-null    datetime64[ns]
 3   Hora             684 non-null    object        
 4   Lugar_Del_Hecho  684 non-null    object        
 5   Tipo_De_Calle    684 non-null    object        
 6   Comuna           684 non-null    int64         
 7   Pos_X            684 non-null    float64       
 8   Pos_Y            684 non-null    float64       
 9   Victima          684 non-null    object        
 10  Acusado          684 non-null    object        
 11  Barrios          684 non-null    object        
dtypes: datetime64[ns](1), float64(2), int64(2), object(7)
memory usage: 69.5+ KB


In [77]:
df_homicidios.info()

<class 'pandas.core.frame.DataFrame'>
Index: 684 entries, 0 to 695
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Id               684 non-null    object        
 1   N_Victimas       684 non-null    int64         
 2   Fecha            684 non-null    datetime64[ns]
 3   Hora             684 non-null    object        
 4   Lugar_Del_Hecho  684 non-null    object        
 5   Tipo_De_Calle    684 non-null    object        
 6   Comuna           684 non-null    int64         
 7   Pos_X            684 non-null    float64       
 8   Pos_Y            684 non-null    float64       
 9   Victima          684 non-null    object        
 10  Acusado          684 non-null    object        
 11  Barrios          684 non-null    object        
dtypes: datetime64[ns](1), float64(2), int64(2), object(7)
memory usage: 69.5+ KB


# Victimas

Importamos el df

In [78]:
df_victimas = pd.read_excel("bases/homicidios.xlsx", sheet_name=2)
df_victimas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID_hecho             717 non-null    object        
 1   FECHA                717 non-null    datetime64[ns]
 2   AAAA                 717 non-null    int64         
 3   MM                   717 non-null    int64         
 4   DD                   717 non-null    int64         
 5   ROL                  717 non-null    object        
 6   VICTIMA              717 non-null    object        
 7   SEXO                 717 non-null    object        
 8   EDAD                 717 non-null    object        
 9   FECHA_FALLECIMIENTO  717 non-null    object        
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 56.1+ KB


seleccionamos unicamente las columnas de interés

In [79]:

columnas_deseadas = ["ID_hecho", "ROL", "SEXO", "EDAD", "FECHA_FALLECIMIENTO"]
df_victimas = df_victimas[columnas_deseadas]

# Muestra el DataFrame resultante
df_victimas

ID_hecho                   ROL       SEXO EDAD  FECHA_FALLECIMIENTO
0    2016-0001             CONDUCTOR  MASCULINO   19  2016-01-01 00:00:00
1    2016-0002             CONDUCTOR  MASCULINO   70  2016-01-02 00:00:00
2    2016-0003             CONDUCTOR  MASCULINO   30  2016-01-03 00:00:00
3    2016-0004             CONDUCTOR  MASCULINO   18                   SD
4    2016-0005             CONDUCTOR  MASCULINO   29  2016-02-01 00:00:00
..         ...                   ...        ...  ...                  ...
712  2021-0092                PEATON   FEMENINO   50  2021-12-12 00:00:00
713  2021-0093  PASAJERO_ACOMPAÑANTE   FEMENINO   18  2021-12-18 00:00:00
714  2021-0094  PASAJERO_ACOMPAÑANTE   FEMENINO   43  2021-12-20 00:00:00
715  2021-0095             CONDUCTOR  MASCULINO   27  2022-01-02 00:00:00
716  2021-0096             CONDUCTOR  MASCULINO   60  2021-12-20 00:00:00

[717 rows x 5 columns]

Normalizamos los nombres

In [80]:
fp.format_nombres(df_victimas)
df_victimas.rename(columns={"Id_Hecho": "Id"}, inplace=True)
df_victimas

Id                   Rol       Sexo Edad  Fecha_Fallecimiento
0    2016-0001             CONDUCTOR  MASCULINO   19  2016-01-01 00:00:00
1    2016-0002             CONDUCTOR  MASCULINO   70  2016-01-02 00:00:00
2    2016-0003             CONDUCTOR  MASCULINO   30  2016-01-03 00:00:00
3    2016-0004             CONDUCTOR  MASCULINO   18                   SD
4    2016-0005             CONDUCTOR  MASCULINO   29  2016-02-01 00:00:00
..         ...                   ...        ...  ...                  ...
712  2021-0092                PEATON   FEMENINO   50  2021-12-12 00:00:00
713  2021-0093  PASAJERO_ACOMPAÑANTE   FEMENINO   18  2021-12-18 00:00:00
714  2021-0094  PASAJERO_ACOMPAÑANTE   FEMENINO   43  2021-12-20 00:00:00
715  2021-0095             CONDUCTOR  MASCULINO   27  2022-01-02 00:00:00
716  2021-0096             CONDUCTOR  MASCULINO   60  2021-12-20 00:00:00

[717 rows x 5 columns]

Exploramos el df

In [81]:
fp.tipos_cols(df_victimas)

Columna                Tipo de Valor  Cantidad  \
0                    Id                <class 'str'>       717   
1                    Id                          NaN         0   
2                   Rol                <class 'str'>       717   
3                   Rol                          NaN         0   
4                  Sexo                <class 'str'>       717   
5                  Sexo                          NaN         0   
6                  Edad                <class 'int'>       664   
7                  Edad                <class 'str'>        53   
8                  Edad                          NaN         0   
9   Fecha_Fallecimiento  <class 'datetime.datetime'>       648   
10  Fecha_Fallecimiento                <class 'str'>        69   
11  Fecha_Fallecimiento                          NaN         0   

   Más de un tipo de valor (no NaN)  
0                                No  
1                                No  
2                                No  
3                                No  
4                                No  
5                                No  
6                                Sí  
7                                Sí  
8                                Sí  
9                                Sí  
10                               Sí  
11                               Sí

No hay valores nulos pero Debemos homogenizar los tipos de datos

## 1. Id

In [82]:
fp.unicos_col(df_victimas["Id"])


Cantidad total de observaciones: 717


tipo  conteo                                            valores
0  <class 'str'>     696  [2016-0001, 2016-0002, 2016-0003, 2016-0004, 2...

In [83]:
fp.frecuencia_unicos(df_victimas["Id"])

Frecuencia de Valores Únicos:
Id
2017-0035    3
2017-0050    2
2016-0041    2
2018-0015    2
2020-0063    2
            ..
2017-0115    1
2017-0116    1
2017-0118    1
2017-0120    1
2021-0096    1
Name: count, Length: 696, dtype: int64


## 2. Rol

In [84]:
fp.unicos_col(df_victimas["Rol"])


Cantidad total de observaciones: 717


tipo  conteo                                            valores
0  <class 'str'>       5  [CONDUCTOR, PASAJERO_ACOMPAÑANTE, PEATON, SD, ...

In [85]:
fp.frecuencia_unicos(df_victimas["Rol"])

Frecuencia de Valores Únicos:
Rol
CONDUCTOR               330
PEATON                  267
PASAJERO_ACOMPAÑANTE     80
CICLISTA                 29
SD                       11
Name: count, dtype: int64


## Sexo

In [86]:
fp.unicos_col(df_victimas["Sexo"])


Cantidad total de observaciones: 717


tipo  conteo                    valores
0  <class 'str'>       3  [MASCULINO, FEMENINO, SD]

In [87]:
fp.frecuencia_unicos(df_victimas["Sexo"])

Frecuencia de Valores Únicos:
Sexo
MASCULINO    545
FEMENINO     166
SD             6
Name: count, dtype: int64


## Edad

In [88]:
fp.unicos_col(df_victimas["Edad"])


Cantidad total de observaciones: 717


tipo  conteo                                            valores
0  <class 'int'>      85  [19, 70, 30, 18, 29, 22, 16, 59, 65, 34, 41, 5...
1  <class 'str'>       1                                               [SD]

Eliminamos ese valor conviertiendo a entero y luego imputando la media

In [89]:
df_victimas["Edad"] = pd.to_numeric(df_victimas["Edad"], errors='coerce', downcast='integer')

media_edades = df_victimas["Edad"].mean()
df_victimas["Edad"] = df_victimas["Edad"].fillna(media_edades).astype(int)



In [90]:
fp.frecuencia_unicos(df_victimas["Edad"])

Frecuencia de Valores Únicos:
Edad
42    60
30    28
29    24
23    24
27    20
      ..
11     1
13     1
85     1
7      1
88     1
Name: count, Length: 85, dtype: int64


Quedó ya correcta

## Fecha de fallecimiento

In [91]:
fp.unicos_col(df_victimas["Fecha_Fallecimiento"])


Cantidad total de observaciones: 717


tipo  conteo  \
0  <class 'datetime.datetime'>     561   
1                <class 'str'>       2   

                                             valores  
0  [2016-01-01 00:00:00, 2016-01-02 00:00:00, 201...  
1                                   [SD, 26/03/2019]

cambiamos los / por -

In [92]:
df_victimas["Fecha_Fallecimiento"] = df_victimas["Fecha_Fallecimiento"].apply(lambda fecha: fecha.replace("/", "-") if isinstance(fecha, str) else fecha)


In [93]:
fp.unicos_col(df_victimas["Fecha_Fallecimiento"])


Cantidad total de observaciones: 717


tipo  conteo  \
0  <class 'datetime.datetime'>     561   
1                <class 'str'>       2   

                                             valores  
0  [2016-01-01 00:00:00, 2016-01-02 00:00:00, 201...  
1                                   [SD, 26-03-2019]

cambiamos a datetime

In [94]:
df_victimas["Fecha_Fallecimiento"] = pd.to_datetime(df_victimas["Fecha_Fallecimiento"], errors='coerce')

In [95]:
fp.unicos_col(df_victimas["Fecha_Fallecimiento"])


Cantidad total de observaciones: 717


tipo  conteo  \
0  <class 'pandas._libs.tslibs.timestamps.Timesta...     562   
1      <class 'pandas._libs.tslibs.nattype.NaTType'>       1   

                                             valores  
0  [2016-01-01 00:00:00, 2016-01-02 00:00:00, 201...  
1                                              [NaT]

Vamos a mantener los nulos para rellenarlos con la fecha del incidente

In [96]:
fp.frecuencia_unicos(df_victimas["Fecha_Fallecimiento"])

Frecuencia de Valores Únicos:
Fecha_Fallecimiento
2017-02-26    3
2019-12-18    3
2018-08-20    3
2020-12-20    3
2018-04-27    3
             ..
2018-01-19    1
2018-01-18    1
2018-01-14    1
2018-01-12    1
2022-01-02    1
Name: count, Length: 562, dtype: int64


## final victimas

In [97]:
fp.tipos_cols(df_victimas)

Columna                                      Tipo de Valor  \
0                    Id                                      <class 'str'>   
1                    Id                                                NaN   
2                   Rol                                      <class 'str'>   
3                   Rol                                                NaN   
4                  Sexo                                      <class 'str'>   
5                  Sexo                                                NaN   
6                  Edad                                      <class 'int'>   
7                  Edad                                                NaN   
8   Fecha_Fallecimiento  <class 'pandas._libs.tslibs.timestamps.Timesta...   
9   Fecha_Fallecimiento      <class 'pandas._libs.tslibs.nattype.NaTType'>   
10  Fecha_Fallecimiento                                                NaN   

    Cantidad Más de un tipo de valor (no NaN)  
0        717                               No  
1          0                               No  
2        717                               No  
3          0                               No  
4        717                               No  
5          0                               No  
6        717                               No  
7          0                               No  
8        649                               Sí  
9         68                               Sí  
10        68                               Sí

# Merge

Vamos a fusionar

In [98]:
df_siniestros = pd.merge(df_homicidios, df_victimas, on="Id", how="left")
df_siniestros

Id  N_Victimas      Fecha      Hora  \
0    2016-0001           1 2016-01-01  04:00:00   
1    2016-0002           1 2016-01-02  01:15:00   
2    2016-0003           1 2016-01-03  07:00:00   
3    2016-0004           1 2016-01-10  00:00:00   
4    2016-0005           1 2016-01-21  05:20:00   
..         ...         ...        ...       ...   
699  2021-0093           1 2021-12-13  17:10:00   
700  2021-0094           1 2021-12-20  01:10:00   
701  2021-0095           1 2021-12-30  00:43:00   
702  2021-0096           1 2021-12-15  10:30:00   
703  2021-0097           1 2021-11-18  06:10:00   

                               Lugar_Del_Hecho Tipo_De_Calle  Comuna  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA       8   
1            AV GRAL PAZ  Y AV DE LOS CORRALES       AVENIDA       9   
2                           AV ENTRE RIOS 2034       AVENIDA       1   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA       8   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA       1   
..                                         ...           ...     ...   
699                          AV. RIESTRA Y MOM       AVENIDA       7   
700                    AU DELLEPIANE Y LACARRA     AUTOPISTA       9   
701                        AV. GAONA Y TERRADA       AVENIDA      11   
702                         AV. EVA PERON 4071       AVENIDA       9   
703                    PADRE CARLOS MUJICA 709         CALLE       1   

         Pos_X      Pos_Y    Victima    Acusado  \
0   -58.475340 -34.687570       Moto       Auto   
1   -58.508775 -34.669777       Auto  Pasajeros   
2   -58.390403 -34.631894       Moto       Auto   
3   -58.465039 -34.680930       Moto         Sd   
4   -58.387183 -34.622466       Moto  Pasajeros   
..         ...        ...        ...        ...   
699 -58.433538 -34.645616       Moto       Auto   
700 -58.467398 -34.651178       Moto       Auto   
701 -58.472934 -34.619847       Moto     Cargas   
702 -58.470668 -34.650217       Auto     Cargas   
703 -58.379762 -34.586796  Bicicleta       Auto   

                                               Barrios                   Rol  \
0       [Villa Soldati, Villa Riachuelo, Villa Lugano]             CONDUCTOR   
1              [Liniers, Mataderos, Parque Avellaneda]             CONDUCTOR   
2    [Retiro, San Nicolás, Puerto Madero, San Telmo...             CONDUCTOR   
3       [Villa Soldati, Villa Riachuelo, Villa Lugano]             CONDUCTOR   
4    [Retiro, San Nicolás, Puerto Madero, San Telmo...             CONDUCTOR   
..                                                 ...                   ...   
699                         [Flores, Parque Chacabuco]  PASAJERO_ACOMPAÑANTE   
700            [Liniers, Mataderos, Parque Avellaneda]  PASAJERO_ACOMPAÑANTE   
701  [Villa General Mitre, Villa Devoto, Villa del ...             CONDUCTOR   
702            [Liniers, Mataderos, Parque Avellaneda]             CONDUCTOR   
703  [Retiro, San Nicolás, Puerto Madero, San Telmo...              CICLISTA   

          Sexo  Edad Fecha_Fallecimiento  
0    MASCULINO    19          2016-01-01  
1    MASCULINO    70          2016-01-02  
2    MASCULINO    30          2016-01-03  
3    MASCULINO    18                 NaT  
4    MASCULINO    29          2016-02-01  
..         ...   ...                 ...  
699   FEMENINO    18          2021-12-18  
700   FEMENINO    43          2021-12-20  
701  MASCULINO    27          2022-01-02  
702  MASCULINO    60          2021-12-20  
703  MASCULINO    53          2021-11-19  

[704 rows x 16 columns]

Solo queda usar la fecha como fecha de fallecimiento allí donde Fecha_Fallecimiento tiene valores nulos

In [99]:
df_siniestros["Fecha_Fallecimiento"] = df_siniestros.apply(lambda row: row["Fecha"] if pd.isnull(row["Fecha_Fallecimiento"]) else row["Fecha_Fallecimiento"], axis=1)


In [100]:
df_siniestros = df_siniestros.rename(columns={"Fecha": "Fecha_Siniestro"})


creamos una variable booleana para usarla más adelante, para indicar el semestre

In [101]:
df_siniestros["Semestre"] = df_siniestros["Fecha_Siniestro"].dt.month.apply(lambda month: 1 if month <= 6 else 2)


Extracción de dia mes y año en fechas

In [102]:
#Hora
df_siniestros['Hora_Siniestro'] = df_siniestros['Hora'].apply(lambda x: x.hour)

#Dia mes y año de siniestros
df_siniestros['Anio_Siniestro'] = df_siniestros['Fecha_Siniestro'].dt.year
df_siniestros['Mes_Siniestro'] = df_siniestros['Fecha_Siniestro'].dt.month
df_siniestros['Dia_Siniestro'] = df_siniestros['Fecha_Siniestro'].dt.day

#Dia mes y año de fallecimientos
df_siniestros['Anio_Fallecimiento'] = df_siniestros['Fecha_Fallecimiento'].dt.year
df_siniestros['Mes_Fallecimiento'] = df_siniestros['Fecha_Fallecimiento'].dt.month
df_siniestros['Dia_Fallecimiento'] = df_siniestros['Fecha_Fallecimiento'].dt.day

obtengamos el dia de la semana en que ocurrieron los eventos

In [103]:
locale.setlocale(locale.LC_TIME, 'es_ES.UTF-8')
df_siniestros['Dia_Semana_Siniestro'] = df_siniestros['Fecha_Siniestro'].dt.strftime('%A')
df_siniestros['Dia_Semana_Fallecimiento'] = df_siniestros['Fecha_Fallecimiento'].dt.strftime('%A')


Veamos el df final

In [104]:
fp.tipos_cols(df_siniestros)

Columna  \
0                         Id   
1                         Id   
2                 N_Victimas   
3                 N_Victimas   
4            Fecha_Siniestro   
5            Fecha_Siniestro   
6                       Hora   
7                       Hora   
8            Lugar_Del_Hecho   
9            Lugar_Del_Hecho   
10             Tipo_De_Calle   
11             Tipo_De_Calle   
12                    Comuna   
13                    Comuna   
14                     Pos_X   
15                     Pos_X   
16                     Pos_Y   
17                     Pos_Y   
18                   Victima   
19                   Victima   
20                   Acusado   
21                   Acusado   
22                   Barrios   
23                   Barrios   
24                       Rol   
25                       Rol   
26                      Sexo   
27                      Sexo   
28                      Edad   
29                      Edad   
30       Fecha_Fallecimiento   
31       Fecha_Fallecimiento   
32                  Semestre   
33                  Semestre   
34            Hora_Siniestro   
35            Hora_Siniestro   
36            Anio_Siniestro   
37            Anio_Siniestro   
38             Mes_Siniestro   
39             Mes_Siniestro   
40             Dia_Siniestro   
41             Dia_Siniestro   
42        Anio_Fallecimiento   
43        Anio_Fallecimiento   
44         Mes_Fallecimiento   
45         Mes_Fallecimiento   
46         Dia_Fallecimiento   
47         Dia_Fallecimiento   
48      Dia_Semana_Siniestro   
49      Dia_Semana_Siniestro   
50  Dia_Semana_Fallecimiento   
51  Dia_Semana_Fallecimiento   

                                        Tipo de Valor  Cantidad  \
0                                       <class 'str'>       704   
1                                                 NaN         0   
2                                       <class 'int'>       704   
3                                                 NaN         0   
4   <class 'pandas._libs.tslibs.timestamps.Timesta...       704   
5                                                 NaN         0   
6                             <class 'datetime.time'>       704   
7                                                 NaN         0   
8                                       <class 'str'>       704   
9                                                 NaN         0   
10                                      <class 'str'>       704   
11                                                NaN         0   
12                                      <class 'int'>       704   
13                                                NaN         0   
14                                    <class 'float'>       704   
15                                                NaN         0   
16                                    <class 'float'>       704   
17                                                NaN         0   
18                                      <class 'str'>       704   
19                                                NaN         0   
20                                      <class 'str'>       704   
21                                                NaN         0   
22                                     <class 'list'>       704   
23                                                NaN         0   
24                                      <class 'str'>       704   
25                                                NaN         0   
26                                      <class 'str'>       704   
27                                                NaN         0   
28                                      <class 'int'>       704   
29                                                NaN         0   
30  <class 'pandas._libs.tslibs.timestamps.Timesta...       704   
31                                                NaN         0   
32                                      <class 'int'>       704   
33                                            

In [105]:
df_siniestros

Id  N_Victimas Fecha_Siniestro      Hora  \
0    2016-0001           1      2016-01-01  04:00:00   
1    2016-0002           1      2016-01-02  01:15:00   
2    2016-0003           1      2016-01-03  07:00:00   
3    2016-0004           1      2016-01-10  00:00:00   
4    2016-0005           1      2016-01-21  05:20:00   
..         ...         ...             ...       ...   
699  2021-0093           1      2021-12-13  17:10:00   
700  2021-0094           1      2021-12-20  01:10:00   
701  2021-0095           1      2021-12-30  00:43:00   
702  2021-0096           1      2021-12-15  10:30:00   
703  2021-0097           1      2021-11-18  06:10:00   

                               Lugar_Del_Hecho Tipo_De_Calle  Comuna  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA       8   
1            AV GRAL PAZ  Y AV DE LOS CORRALES       AVENIDA       9   
2                           AV ENTRE RIOS 2034       AVENIDA       1   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA       8   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA       1   
..                                         ...           ...     ...   
699                          AV. RIESTRA Y MOM       AVENIDA       7   
700                    AU DELLEPIANE Y LACARRA     AUTOPISTA       9   
701                        AV. GAONA Y TERRADA       AVENIDA      11   
702                         AV. EVA PERON 4071       AVENIDA       9   
703                    PADRE CARLOS MUJICA 709         CALLE       1   

         Pos_X      Pos_Y    Victima  ... Semestre Hora_Siniestro  \
0   -58.475340 -34.687570       Moto  ...        1              4   
1   -58.508775 -34.669777       Auto  ...        1              1   
2   -58.390403 -34.631894       Moto  ...        1              7   
3   -58.465039 -34.680930       Moto  ...        1              0   
4   -58.387183 -34.622466       Moto  ...        1              5   
..         ...        ...        ...  ...      ...            ...   
699 -58.433538 -34.645616       Moto  ...        2             17   
700 -58.467398 -34.651178       Moto  ...        2              1   
701 -58.472934 -34.619847       Moto  ...        2              0   
702 -58.470668 -34.650217       Auto  ...        2             10   
703 -58.379762 -34.586796  Bicicleta  ...        2              6   

    Anio_Siniestro Mes_Siniestro  Dia_Siniestro Anio_Fallecimiento  \
0             2016             1              1               2016   
1             2016             1              2               2016   
2             2016             1              3               2016   
3             2016             1             10               2016   
4             2016             1             21               2016   
..             ...           ...            ...                ...   
699           2021            12             13               2021   
700           2021            12             20               2021   
701           2021            12             30               2022   
702           2021            12             15               2021   
703           2021            11             18               2021   

     Mes_Fallecimiento  Dia_Fallecimiento  Dia_Semana_Siniestro  \
0                    1                  1               viernes   
1                    1                  2               sábado   
2                    1                  3               domingo   
3                    1                 10               domingo   
4                    2                  1                jueves   
..                 ...                ...                   ...   
699                 12                 18                 lunes   
700                 12                 20                 lunes   
701                  1                  2                jueves   
702                 12                 20            miércoles   
703                 11                 19                jueves   


Lo exportamos

In [106]:
ruta_archivo_parquet = 'bases/siniestros.parquet'

# Exportar el DataFrame a un archivo Parquet
df_siniestros.to_parquet(ruta_archivo_parquet, index=False)


# Guardar el DataFrame como un archivo CSV
df_siniestros.to_csv('bases/siniestros.csv', index=False, encoding='utf-8')


Importamos y revisamos que todo esté bien

In [107]:
df_siniestros_imp = pd.read_parquet(ruta_archivo_parquet)
df_siniestros_imp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 704 entries, 0 to 703
Data columns (total 26 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Id                        704 non-null    object        
 1   N_Victimas                704 non-null    int64         
 2   Fecha_Siniestro           704 non-null    datetime64[ns]
 3   Hora                      704 non-null    object        
 4   Lugar_Del_Hecho           704 non-null    object        
 5   Tipo_De_Calle             704 non-null    object        
 6   Comuna                    704 non-null    int64         
 7   Pos_X                     704 non-null    float64       
 8   Pos_Y                     704 non-null    float64       
 9   Victima                   704 non-null    object        
 10  Acusado                   704 non-null    object        
 11  Barrios                   704 non-null    object        
 12  Rol                   